packages/skygear-core/src/client.ts

```diff
@@ -857,4 +884,90 @@ export abstract class BaseAPIClient {
       json: {},
     });
   }
+
+  /**
+   * @internal
+   */
+  async _fetchOIDCRequest(url: string, init?: RequestInit): Promise<any> {
+    const resp = await this._fetch(url, init);
+    let jsonBody;
+    try {
+      jsonBody = await resp.json();
+    } catch {
+      throw new SkygearError(
+        "failed to decode response JSON",
+        "InternalError",
+        "UnexpectedError",
+        {
+          status_code: resp.status,
+        }
+      );
+    }
+    if (resp.status === 200) {
+      return jsonBody;
+    }
+    const oauthError: OAuthError = {
+      error: jsonBody["error"],
+      error_description: jsonBody["error_description"],
+    };
+    throw oauthError;
+  }
+
+  /**
+   * @internal
+   */
+  async _fetchOIDCConfiguration(): Promise<_OIDCConfiguration> {
+    if (!this.config) {
+      this.config = (await this._fetchOIDCRequest(
+        `${this.authEndpoint}/.well-known/openid-configuration`
+      )) as _OIDCConfiguration;
+    }
+    return this.config;
+  }
+
+  /**
+   * @internal
+   */
+  async _oidcTokenRequest(req: _OIDCTokenRequest): Promise<_OIDCTokenResponse> {
+    const config = await this._fetchOIDCConfiguration();
+    const query: [string, string][] = [];
+    query.push(["grant_type", req.grant_type]);
+    query.push(["client_id", req.client_id]);
+    if (req.code) {
+      query.push(["code", req.code]);
+    }
+    if (req.redirect_uri) {
+      query.push(["redirect_uri", req.redirect_uri]);
+    }
+    if (req.code_verifier) {
+      query.push(["code_verifier", req.code_verifier]);
+    }
+    if (req.refresh_token) {
+      query.push(["refresh_token", req.refresh_token]);
+    }
+    return this._fetchOIDCRequest(config.token_endpoint, {
+      method: "POST",
+      headers: {
+        "content-type": "application/x-www-form-urlencoded",
+      },
+      body: encodeQuery(query).substring(1),
+    });
+  }
+
+  /**
+   * @internal
+   */
+  async _oidcUserInfoRequest(accessToken?: string): Promise<AuthResponse> {
+    const headers: { [name: string]: string } = {};
+    if (accessToken) {
+      headers["authorization"] = `bearer ${accessToken}`;
+    }
+    const config = await this._fetchOIDCConfiguration();
+    const userinfo = await this._fetchOIDCRequest(config.userinfo_endpoint, {
+      method: "GET",
+      headers: headers,
```

Comment:
Add 

```typescript
mode: "cors",
credentials: "include"
```

because we sometimes expect cookie to be included for the userinfo endpoint